asmcomp/linearize.ml

```diff
@@ -294,14 +295,19 @@ let rec linear i n =
       loop (add_branch lbl n1) !try_depth
   | Itrywith(body, handler) ->
       let (lbl_join, n1) = get_label (linear i.Mach.next n) in
+      let (lbl_handler, n2) =
+        get_label (cons_instr Lentertrap (linear handler n1))
+      in
       incr try_depth;
       assert (i.Mach.arg = [| |] || Config.spacetime);
-      let (lbl_body, n2) =
-        get_label (instr_cons Lpushtrap i.Mach.arg [| |]
-                    (linear body (cons_instr Lpoptrap n1))) in
-      decr try_depth;
-      instr_cons (Lsetuptrap lbl_body) i.Mach.arg [| |]
-        (linear handler (add_branch lbl_join n2))
+      let n3 = cons_instr (Lpushtrap { lbl_handler })
+                 (linear body
+                    (cons_instr
+                       Lpoptrap
+                       (add_branch lbl_join n2))) in
+      decr try_depth; (* does it need to come before the use of n2? *)
```

Comment:
There is a question here to be answered.

Reply:
The call to `linear body` must be between the `incr try_depth` and `decr try_depth` statements, and the other recursive calls to `linear` must not.
There is no other constraint, so the code is fine. I'll remove the comment.